### PR TITLE
[Model] LTD

### PR DESCRIPTION
**Pull request by honey0219**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/ltd/distill.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    msg = tlx.gather(teacher_softmax, edge_index_no_self_loops[1])
+    nei_logits_sum = unsorted_segment_sum(msg, edge_index_no_self_loops[0], graph.num_nodes)
+    nei_num = unsorted_segment_sum(tlx.ops.ones_like(edge_index_no_self_loops[0], dtype=torch.float),
+                                   edge_index_no_self_loops[0],
+                                   graph.num_nodes)
+    nei_probability = tlx.ops.transpose(tlx.ops.divide(tlx.ops.transpose(nei_logits_sum), nei_num))
+    nei_entropy = -tlx.ops.reduce_sum(nei_probability * tlx.ops.log(nei_probability), axis=1)
+    for i in range(len(nei_entropy)):
+        if nei_entropy[i] != nei_entropy[i]:
+            nei_entropy[i] = 0.0001
```
Comment (gyzhou2000) [inline examples/ltd/distill.py:154]: 计算损失函数的这部分代码可以考虑整理成为一个函数进行调用

File: examples/ltd/distill.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    t_model = MLP(num_layers=2, in_channels=configs['num_classes'] + 2, hidden_channels=64, out_channels=1, dropout=0.6,
+                  norm=None, act=tlx.nn.ReLU())
+
+    best = 0
+    cnt = 0
+    epoch = 1
+    while epoch < configs['max_epoch']:
+        acc_val, train_loss = distill_train(epoch, model, configs, graph, nei_entropy, t_model,
+                                            teacher_logits)
+        if epoch > 0:
```
Comment (gyzhou2000) [inline examples/ltd/distill.py:165]: `epoch` 这个值恒大于0吧，似乎没起作用

File: examples/ltd/distill.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    model.set_eval()
+    edge_index, _ = add_self_loops(graph.edge_index, num_nodes=graph.num_nodes)
+    if configs['student'] in ['GCN', 'LogReg', 'MLP']:
+        logits = model(graph.x, edge_index, None, graph.num_nodes)
+    if configs['student'] in ['APPNP']:
+        logits = model(graph.x)
+    elif configs['student'] == 'GAT':
+        logits = model(graph.x, edge_index, graph.num_nodes)
+    elif configs['student'] in ['GraphSAGE', 'SGC']:
+        logits = model(graph.x)
```
Comment (gyzhou2000) [inline examples/ltd/distill.py:194]: `train_conf.yaml` 文件中只给出了 `GCN` 和 `GAT` 的超参数，可以考虑给出其他模型的超参数

File: examples/ltd/distill.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+                         hidden_dim=configs['hidden_dim'],
+                         num_class=configs['num_classes'],
+                         heads=configs['heads'],
+                         drop_rate=configs['drop_rate'],
+                         num_layers=configs['num_layers'],
+                         name="GAT",
+                         )
+    else:
+        raise ValueError(f'Undefined Model.')
+    return model
```
Comment (gyzhou2000) [inline examples/ltd/distill.py:54]: 这个函数看起来放在 `ltd_trainer.py` 文件中更合适，模型的选择和distill过程无关的，可以考虑放在 `ltd_traier.py` 文件中

File: examples/ltd/distill.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                                temparature)
+    teacher_logits_t = tlx.ops.transpose(
+        tlx.ops.divide(tlx.ops.transpose(teacher_logits), tlx.ops.squeeze(temparature, axis=1)))
+    teacher_softmax = tlx.ops.softmax(teacher_logits_t, axis=1)
+    student_hard = tlx.ops.softmax(logits, axis=1)
+    labels_one_hot = tlx.nn.OneHot(depth=configs['num_classes'])(graph.y)
+    hard_loss = -tlx.ops.reduce_sum(
+        (labels_one_hot[configs['train_mask']] + 1e-6) * tlx.ops.log(student_hard[configs['train_mask']] + 1e-6))
+    soft_loss = -tlx.ops.reduce_sum((teacher_softmax + 1e-6) * tlx.ops.log(student_hard + 1e-6))
+    train_loss = soft_loss + configs['lam'] * hard_loss
```
Comment (gyzhou2000) [inline examples/ltd/distill.py:92]: 计算损失函数的这一部分可以考虑写在一个函数中

File: examples/ltd/distill.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        edge_index, _ = add_self_loops(graph.edge_index, num_nodes=graph.num_nodes)
+        if configs['student'] == 'GCN':
+            logits = model(graph.x, edge_index, None, graph.num_nodes)
+        elif configs['student'] == 'GAT':
+            logits = model(graph.x, edge_index, graph.num_nodes)
+        logits = layer_normalize(logits)
+        student_hard = tlx.ops.softmax(logits, axis=1)
+        labels_one_hot = tlx.nn.OneHot(depth=configs['num_classes'])(graph.y)
+        t_train_loss = -tlx.ops.reduce_sum(
+            (labels_one_hot[configs['val_mask']] + 1e-6) * tlx.ops.log(student_hard[configs['val_mask']] + 1e-6))
```
Comment (gyzhou2000) [inline examples/ltd/distill.py:111]: 同上

File: examples/ltd/distill.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    if configs['student'] == 'GCN':
+        logits = model(graph.x, edge_index, None, graph.num_nodes)
+    elif configs['student'] == 'GAT':
+        logits = model(graph.x, edge_index, graph.num_nodes)
+    acc_val = accuracy(logits[configs['my_val_mask']], graph.y[configs['my_val_mask']])
+    print(
+        'Epoch %d | acc_train: %.4f| acc_val: %.4f | acc_same: %.4f | hard_loss: %.4f | soft_loss: %.4f| distill_loss: %.4f | evaluate_loss: %.4f' % (
+            epoch, acc_train.item(), acc_val.item(), acc_same.item(), hard_loss.item(), soft_loss.item(),
+            train_loss.item(),
+            t_train_loss.item()))
```
Comment (gyzhou2000) [inline examples/ltd/distill.py:138]: 该变量命名可以考虑用更贴切的名称

File: examples/ltd/distill.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    edge_index_no_self_loops, _ = remove_self_loops(graph.edge_index)
+    msg = tlx.gather(teacher_softmax, edge_index_no_self_loops[1])
+    nei_logits_sum = unsorted_segment_sum(msg, edge_index_no_self_loops[0], graph.num_nodes)
+    nei_num = unsorted_segment_sum(tlx.ops.ones_like(edge_index_no_self_loops[0], dtype=torch.float),
+                                   edge_index_no_self_loops[0],
+                                   graph.num_nodes)
+    nei_probability = tlx.ops.transpose(tlx.ops.divide(tlx.ops.transpose(nei_logits_sum), nei_num))
+    nei_entropy = -tlx.ops.reduce_sum(nei_probability * tlx.ops.log(nei_probability), axis=1)
+    for i in range(len(nei_entropy)):
+        if nei_entropy[i] != nei_entropy[i]:
```
Comment (dddg617) [inline examples/ltd/distill.py:26]: Can you explain this? Under what circumstances can this happen? And why do you set this to "0.0001"?

Comment (honey0219) [reply examples/ltd/distill.py:26]: 我理解的是为了避免nei_entropy里出现NaN，把NaN的地方设成一个很小的数，0.0001是参照论文源码设置的。出现这种情况是因为nei_num中可能会有0值，除零错误会导致NaN吧

**Comment by dddg617** [comment]
I think in the next version, it will be better to support the gradient operation in `TensorLayerX` through this work. @gyzhou2000 . 
The next work @honey0219 needs to do is to provide the document. It doesn't matter if the performance difference is within 3%.

File: examples/ltd/readme.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+| Dataset  | # Nodes | # Edges | # Classes |
+|----------|---------|---------|-----------|
+| Cora     | 2,708   | 10,556  | 7         |
+| Citeseer | 3,327   | 9,228   | 6         |
+| Pubmed   | 19,717  | 88,651  | 3         |
+
+Refer to [Planetoid](https://gammagl.readthedocs.io/en/latest/api/gammagl.datasets.html#gammagl.datasets.Planetoid).
+
+Results
+-------
```
Comment (dddg617) [inline examples/ltd/readme.md:17]: This should be `# Results` instead of the `rst`.
